text/2019-11-13-follower-replication.md

```diff
@@ -64,64 +66,34 @@ Here is a diagram showing how Follower Replication works:
-
-```
-Commission {
-    type  // Only two types: Append or Snapshot
-    target  // The target group member
-    last_index
-    log_term
-}
-```
+We can add a new field into `Message`: `BcastTargets`, which is a list of peer IDs that the message receiver (the delegate) needs to replicate entries to. When a peer become delegate first time, it doesn't know followers' progress, so it appends its last log entry to other peers. After it receives `MsgAppendResponse` from those peers, it can update their progresses.
 
 You can just treat a commission as the metadata of a MsgAppend or MsgSnapshot since the `last_index` and `log_term` are set by the leader according to its progress set.
 
-When the delegate receives a `MsgBroadcast`, it might meet any scenario below:
-
-1. If the message declares that the delegate needs a snapshot, it means that all the group members are requiring snapshots. Every commission type should be Snapshot and the delegate just broadcasts to others.
-2. If the message declares that the delegate needs entries, just like the origin raft protocol, it first tries to append incoming entries to its unstable raft log. And if a conflict ocurrs, it sends a rejecting message to the leader and then the leader will try to pick a new delegate to re-send commissions again.
-3. If step 2 succeeds, the delegate will try to execute all the commissions by gathering correspond entries and sending them to each group member. During this time, some of commissions could be failed due to the stale progress state in the leader node. The delegate will collect all the failed commissions and send them back to the leader to trigger a new broadcast message so that Log Replication is always ongoing.
-
-All the other group members have no idea whether a `MsgAppend` or `MsgSnapshot` comes from the leader or the delegate because the delegate always replaces the `from` of the message with the leader ID when executing a commission. So they finally handle these messages like normal followers.
-
-It’s significant to update the inflight of progress when the corresponding commission is generated and rollback the inflight once the leader receives failed commissions in step3 above, which also guarantees that the actual progress of group members (except the delegate) will not be stale.
+Peers can know `MsgAppend` comes from a delegate or its leader, because an another field `delegate` is added in `Message`. If a peer receives a `MsgAppend` from a delegate, it needs to response both its leader and the delegate, so that both its leader and delegate can update progress for it.
 
 ## Drawbacks
 
-When there is a large group or many groups, the log committing speed could be slow as entries will be sent to delegate first.
+1. Committing speed could be slow if `quorum` must involve peers in different data center.
+2. When handle a Raft message, a hashmap query on `map[peer_id]group_id` is introduced.
 
 ## Alternatives
 
```

Comment:
```suggestion
In the current design, the leader not only let the delegate replicate logs to other peers but also 'delegate's the whole flow control to the delegate. We can benefit from it because of the lighter cross-data-center network transmission between the leader and peers but suffer from some issues described in **Drawbacks** section.
There is an alternative design that the leader doesn't 'delegate' the flow control to the delegate. When the leader sends `BcastTargets` to the delegate, it adds the `last_index` of each member in `BcastTargets` to indicate the peer's progress from the leader's view. In this way, the messaging flow is always controlled by the leader and the `Inflights` in-consistency can be avoided. 
```